src/tiledb/cloud/geospatial/ingestion.py

```diff
@@ -0,0 +1,1405 @@
+
+
+def _wrap_paths(
+    paths: Optional[Union[Sequence[os.PathLike], os.PathLike]]
+) -> Optional[Tuple[os.PathLike, ...]]:
+    if paths is None:
+        return None
+        # Would it make sense to return () here, so it's always a Tuple[PathLike, ...]?
+    if isinstance(paths, (str, bytes)) or not isinstance(paths, Sequence):
+        # Editor's note: the isinstance has to go after the str-or-bytes check
+        # because str and bytes are both Sequences.
+        # (Worse yet, str is a Sequence[str].)
+        return (paths,)
+    return tuple(paths)
+
+
+@attrs.define
+class GeoMetadata:
+    # common properties
+    extents: BoundingBox
+    crs: str = None
+    paths: attrs.field(converter=_wrap_paths, default=None) = None
+    # raster properties
+    dtype: Optional[str] = None
+    band_count: Optional[int] = None
+    res: Optional[Tuple[float, float]] = None
+    block_metadata: Optional[Tuple[GeoBlockMetadata, ...]] = None
+    # point cloud properties
+    scales: Optional[XYZTuple] = None
+    offsets: Optional[XYZTuple] = None
+    # geometry properties
+    geometry_schema: Optional[dict] = None
+
+
+def load_pointcloud_metadata(
+    sources: Iterable[os.PathLike],
+    *,
+    config: Optional[Mapping[str, object]] = None,
+    verbose: bool = False,
+    id: str = "pointcloud_metadata",
+    trace: bool = False,
+    log_uri: Optional[str] = None,
+) -> GeoMetadata:
+    """Return geospatial metadata for a sequence of input point cloud data files
+
+    :param sources: iterator, paths or path to process
+    :param config: dict, configuration to pass on tiledb.VFS
+    :param verbose: bool, enable verbose logging, default is False
+    :param trace: bool, enable trace logging, default is False
+    :param log_uri: Optional[str] = None,
+    :Return: GeoMetadata, a populated GeoMetadata object
+    """
+    if not sources:
+        raise ValueError("Input point cloud datasets required")
+
+    def _fold_in(oper, existing, new):
+        if existing is None:
+            return new
+        return [oper(e) for e in zip(existing, new)]
+
+    with tiledb.scope_ctx(config):
+        vfs = tiledb.VFS()
+        logger = get_logger_wrapper(verbose)
+        with Profiler(array_uri=log_uri, id=id, trace=trace):
+            offsets = None
+            scales = None
+            mins = None
+            maxs = None
+            for f in sources:
+                logger.debug("finding metadata for %r", f)
+                with vfs.open(f, "rb") as src:
+                    # only open header
+                    las = laspy.open(src)
+                    hdr = las.header
+                    mins = _fold_in(min, mins, hdr.mins)
+                    maxs = _fold_in(max, maxs, hdr.maxs)
+                    scales = _fold_in(max, scales, hdr.scales)
+                    offsets = _fold_in(min, offsets, hdr.offsets)
+
+            return GeoMetadata(
+                paths=sources,
```

Comment:
This needs to be `list(sources)`

Reply:
Actually we should convert source to list before we consume the generator. something like:
```paths = list(sources)
for f in paths:
...
...
..
return GeoMetadata(
    paths=paths
...
...
```

Reply:
👍